Index: zmqeventconsumer.cpp
===================================================================
--- tango_812c/lib/cpp/client/zmqeventconsumer.cpp
+++ tango_812c.p1/lib/cpp/client/zmqeventconsumer.cpp
@@ -1879,7 +1892,7 @@
     if (ev_name.find(MODIFIER_DBASE_NO) != string::npos)
 		no_db_dev = true;
 
-    for (loop = 0;loop < env_var_fqdn_prefix.size();loop++)
+    for (loop = 0;loop < env_var_fqdn_prefix.size() + 1;loop++)
     {
 
 //
@@ -1888,10 +1901,10 @@
 
 		string new_tango_host;
 
-		if (no_db_dev == true)
+		if (loop == 0 || no_db_dev == true)
 			new_tango_host = ev_name;
 		else
-			new_tango_host = env_var_fqdn_prefix[loop] + canon_ev_name;
+			new_tango_host = env_var_fqdn_prefix[loop - 1] + canon_ev_name;
 
         ipos = event_callback_map.find(new_tango_host);
